File: backend/src/models/utils/adBuilder.js
const INTEGRATIONS_COLLECTION = 'integrations'

class ADBuilder {
  buildForLLMField(userId, workflowId, fieldPath) {
    return this.build(INTEGRATIONS_COLLECTION, userId, normalizeWorkflowId(workflowId), fieldPath)
  }

  buildForArrayField(userId, workflowId, arrayName, alias, fieldPath) {
    return this.build(INTEGRATIONS_COLLECTION, userId, normalizeWorkflowId(workflowId), arrayName, alias, fieldPath)
  }

  build(...parts) {
    const totalSize = parts.reduce((sum, part) => sum + 4 + Buffer.byteLength(part, 'utf8'), 0)
    const buffer = Buffer.allocUnsafe(totalSize)

    let offset = 0
    for (const part of parts) {
      const partBytes = Buffer.from(part, 'utf8')
      buffer.writeUInt32BE(partBytes.length, offset)
      offset += 4
      partBytes.copy(buffer, offset)
      offset += partBytes.length
    }

    return buffer
  }
}

function normalizeWorkflowId(workflowId) {
  return workflowId || ''
}

export default new ADBuilder()

File: backend-v2/internal/common/encryption/ad_builder.go
package encryption

import (
	"encoding/binary"
)

const (
	IntegrationsCollection = "integrations"
)

type ADBuilder struct{}

func NewADBuilder() *ADBuilder {
	return &ADBuilder{}
}

func (b *ADBuilder) BuildForLLMField(userID, workflowID, fieldPath string) []byte {
	return b.build(IntegrationsCollection, userID, workflowID, fieldPath)
}

func (b *ADBuilder) BuildForArrayField(userID, workflowID, arrayName, alias, fieldPath string) []byte {
	return b.build(IntegrationsCollection, userID, workflowID, arrayName, alias, fieldPath)
}

func (b *ADBuilder) build(parts ...string) []byte {
	totalSize := 0
	for _, part := range parts {
		totalSize += 4 + len(part)
	}

	buffer := make([]byte, 0, totalSize)

	for _, part := range parts {
		partBytes := []byte(part)
		lengthBytes := make([]byte, 4)
		binary.BigEndian.PutUint32(lengthBytes, uint32(len(partBytes)))

		buffer = append(buffer, lengthBytes...)
		buffer = append(buffer, partBytes...)
	}

	return buffer
}
